### PR TITLE
Implement provider

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @inikulin @AlexanderMoskovkin

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+language: node_js
+node_js:
+  - stable
+  - v4
+  - '0.12'
+  - '0.10'
```
Comment (AlexanderMoskovkin) [inline .travis.yml]: I think it's enough to leave just `0.10` and `stable`

File: Gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+gulp.task('build', ['lint', 'clean'], function () {
+    return gulp
+        .src('src/**/*.js')
+        .pipe(babel())
+        .pipe(gulp.dest('lib'));
+});
+
+gulp.task('test-mocha', ['build'], function () {
+    if (!process.env.BROWSERSTACK_USERNAME || !process.env.BROWSERSTACK_ACCESS_KEY)
+        throw new Error('Specify your credentials by using the SAUCE_USERNAME and SAUCE_ACCESS_KEY environment variables to authenticate to SauceLabs.');
```
Comment (AlexanderMoskovkin) [inline Gulpfile.js]: fix the message. It's about `saucelabs` now instead of `browserstack`

File: package.json [hunk 3]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "testcafe-browser-provider-browserstack",
+  "version": "1.0.0",
```
Comment (AlexanderMoskovkin) [inline package.json]: let's start with `0.0.1`, we'll bump the version after we test it.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  ],
+  "scripts": {
+    "test": "gulp test",
+    "publish-please": "publish-please",
+    "prepublish": "publish-please guard"
+  },
+  "keywords": [
+    "testcafe",
+    "browser provider",
+    "plugin"
```
Comment (AlexanderMoskovkin) [inline package.json]: add `browserstack`

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        await this._getConnector();
+
+        this.workers[id]         = await doRequest('https://api.browserstack.com/4/worker', 'POST', capabilities);
+        this.workers[id].started = Date.now();
+    },
+
+    async closeBrowser (id) {
+        var workerTime = Date.now() - this.workers[id].started;
+
+        if (workerTime < 30000) {
```
Comment (AlexanderMoskovkin) [inline src/index.js]: What the magic numbers?

File: package.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  "license": "MIT",
+  "dependencies": {
+    "babel-runtime": "^6.11.6",
+    "browserstack-local": "^1.3.0",
+    "desired-capabilities": "^0.1.0",
+    "jimp": "^0.2.27",
+    "pinkie": "^2.0.4",
+    "request": "^2.79.0",
+    "request-promise": "^4.1.1",
+    "stream-to": "^2.0.0"
```
Comment (AlexanderMoskovkin) [inline package.json]: It seems `request` and `stream-to` are redundant

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,227 @@
+import { stringify as makeQueryString } from 'querystring';
+import Promise from 'pinkie';
+import request from 'request-promise';
+import parseCapabilities from 'desired-capabilities';
+import { Local as BrowserstackConnector } from 'browserstack-local';
+import jimp from 'jimp';
+
+
+const TESTS_TIMEOUT                = 1800;
```
Comment (AlexanderMoskovkin) [inline src/index.js]: I think it can be not enough for some tasks. Maybe we should set it as an env variable optionally?

File: package.json (hunk 6 above)
Comment (AndreyBelym) [inline package.json]: `request` is a peer dependency of `request-promise`, so we have to specify it in `package.json`.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# testcafe-browser-provider-browserstack
+[![Build Status](https://travis-ci.org/DevExpress/testcafe-browser-provider-browserstack.svg)](https://travis-ci.org/DevExpress/testcafe-browser-provider-browserstack)
+
+This is the [TestCafe](http://devexpress.github.io/testcafe) browser provider plugin for the integration with the [BrowserStack Testing Cloud](https://browserstack.com/).
```
Comment (VasilyStrelyaev) [inline README.md]:
```
This plugin integrates [TestCafe](http://devexpress.github.io/testcafe) with the [BrowserStack Testing Cloud](https://browserstack.com/).
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+## Usage
+Before using the provider, save the BrowserStack username and access key to environment variables `BROWSERSTACK_USERNAME` and `BROWSERSTACK_ACCESS_KEY`.
+
+You can determine the available browser aliases by running
+```
+testcafe -b browserstack
+```
+
+When you run tests from the command line, use the alias when specifying browsers:
```
Comment (VasilyStrelyaev) [inline README.md]:
```
If you run tests from the command line, use the alias when specifying browsers:
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+This is the [TestCafe](http://devexpress.github.io/testcafe) browser provider plugin for the integration with the [BrowserStack Testing Cloud](https://browserstack.com/).
+
+## Install
+
+```
+npm install testcafe-browser-provider-browserstack
+```
+
+## Usage
+Before using the provider, save the BrowserStack username and access key to environment variables `BROWSERSTACK_USERNAME` and `BROWSERSTACK_ACCESS_KEY`.
```
Comment (VasilyStrelyaev) [inline README.md]:
```
Before using this plugin, save the BrowserStack username and access key to environment variables `BROWSERSTACK_USERNAME` and `BROWSERSTACK_ACCESS_KEY`.
```

**Comment by AndreyBelym** [comment]
FPR

**Pull request merged**